Standard deviation implementation on Measure-Object

CloudyDino opened the pull request:
## PR Summary

<!-- summarize your PR between here and the checklist -->

This continues the work done by @dfinke on pull request #4969. It removes the rounding before displaying the standard deviation, and addresses comments by @iSazonov and @SteveL-MSFT about removing code due to different parameter sets. It passes Measure-Object.Tests.ps1.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

dfinke committed:
First pass at adding std deviation to measure-object

dfinke committed:
Use A*A instead of Math.Pow. Throw if -Average not specified and fixed comments

dfinke committed:
Added message to the rex file

dfinke committed:
Set Average true if StdDeviation is set

dfinke committed:
Update message

CloudyDino committed:
Addressed comments by @iSazonov and @SteveL-MSFT

CloudyDino committed:
Cast average outside for loop

CloudyDino committed:
Updated Measure-Object.Tests.ps1 to match standard deviation calculations

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -71,6 +71,13 @@ public GenericMeasureInfo()
+
+        /// <summary>
+        ///
+        /// The Standard Deviation of property values
```
Please add final dot in all public comments.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -71,6 +71,13 @@ public GenericMeasureInfo()
         public double? Minimum { get; set; }
+
+        /// <summary>
+        ///
```
Please remove such extra lines.

CloudyDino replied:
We're talking about the empty comment lines like line 76 right?

CloudyDino replied:
Also, do we want to remove lines like line 378 where there is no summary included?

iSazonov replied:
Yes, please remove empty comment lines in your code. We usually put comment right under line where we see a problem.
The empty line in public comment gets into auto-generated documentation and corrupts its formatting.
The 378 line is not your code - you shouldn't correct it until reviewer/maintainer ask you. Usually we open new tracking Issue for changing bad patterns and fix its in separate PR/commit.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -263,6 +279,28 @@ public MeasureObjectCommand()
+        /// <summary>
+        /// Set to true if Standard Deviation is to be returned
+        /// </summary>
+        /// <value></value>
```
Please remove the line.

CloudyDino replied:
Lines like 285 with empty value tags?

iSazonov replied:
Yes, empty value tags don't contain useful information.

vors replied:
remove this `/// <value></value>` line as discussed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/resources/MeasureObjectStrings.resx:
```diff
@@ -126,4 +126,7 @@
   <data name="NonNumericInputObject" xml:space="preserve">
     <value>Input object "{0}" is not numeric.</value>
   </data>
+  <data name="AverageSwitchNotSet" xml:space="preserve">
```
We don't use the `AverageSwitchNotSet`.

iSazonov commented:
I have some questions about the implementation
- should we implement "sample standard deviation" or "corrected sample standard deviation"? I'd expect first
- should we implement "one pass" algorithm without store all values (more cpu load)? I'd vote for "one pass"
- should we throw if StdDeviation=$true and Average=$false? I would consider the presence of individual parameters as a request for output values. These values are nullable. So if we set `-Count` we want to get the count in output. If we set `-Average` we want get the average in output. And so on. It doesn't affect what we do internally. So `-Average` can assume that we do count.

iSazonov commented:
@tylerl0706 I suppose @vors is leading this. I believe he "translate" strange requests from maintainers and community 😄 If students have questions - feel free to ask - you are protected by Hack Illinois 😄

vors commented:
Need to do an experiment to prove this, but my gut feeling is that one-pass implementation will suffer considerably from floating point arithmetic.
I think that memory allocation is ok, if user explicitly asked for standard deviation information. PowerShell often operates with large in-memory collections and this place hardly can be a bottleneck.

iSazonov commented:
My concern is that PowerShell is too eats in all subsystems. I think we _should think_ about being more effective every time.
For the feature we haven't source scenario so we can only do general reasoning. What I have is thousands of objects for which I can count the deviation. Not too much. There's one more task out of millions of objects but this is a database and I can use SQL instead PowerShell.

vors commented:
I think original concerns about the arithmetic precision are invalid.

I run this code to calculate `sd(1:100000)^2`

```powershell
$s2 = 0.0
$x = 0.0
$n = 100000
$data = 1..$n
foreach ($i in 1..$n) {
    $xPrev = $x 
    $x = $xPrev + ($data[$i-1] - $xPrev) / $i
    if ($i -gt 1) {
        $s2 = ($i - 2) / ($i - 1) * $s2 + ($data[$i-1] - $xPrev)*($data[$i-1] - $xPrev) / $i
    }
}
$s2
```

The output is `833341666.666671` compare to the `833341667` that R produces (R implements the precision estimation better and rounds it).
So I agree with @iSazonov : we should implement the iterative approach.
I took formulas from wikipedia https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

@CloudyDino would you be interested in rewriting the algorithm to the iterative approach?

CloudyDino commented:
Yeah, I'd love to give this another try with the iterative method! A bit new to GitHub, would I just delete this branch of mine and make a new one to implement it? Or should I just make a new branch without deleting this one? @vors

tylerl0706 commented:
@CloudyDino All you have to do is make changes to the branch you're already working off of and push them up. Github will notice that you made changes and those changes will be automatically reflected here 😄

vors commented:
Awesome! Yes, what @tylerl0706 said: push to the same brunch and PR will be updated. More details: https://stackoverflow.com/questions/9790448/how-to-update-a-pull-request-from-forked-repo

CloudyDino committed:
Change Standard Deviation calculation to be iterative

CloudyDino commented:
I haven't changed the tester file at all and I'm not sure if I'm failing the test if I wrote my code wrong or if the iterative method needs that extra wiggle room. I've been trying to figure this out for a bit. Am I failing anything besides the tests in PowerShell/test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1 ?

vors commented:
@CloudyDino to find out what exactly failed, you can click the "Details" links next to the red crosses. And then dig thru logs a little bit. In our case we can see the following:

https://ci.appveyor.com/project/PowerShell/powershell/build/v6.1.0-preview.8336/tests

```
Expected .00000000000001 to be less than the actual value, but got 0.0280303239036055.
```

Since `$testObject = 1,3,4` we really should not get that big of the difference in the methods. I think the implementation has a bug. Let me take a look.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -708,8 +739,20 @@ private void AnalyzeString(string strValue, Statistics stat)
+                stat.stdDeviation *= countMinusOne;
+                stat.stdDeviation += (numValue - avgPrevious) * (numValue - avg);
+                stat.stdDeviation /= stat.count;
+            }
```
I think the implementation is slightly incorrect: if we have only 1 entry you algorithm doesn't output 0, like it should. I also expected to see a `(count - 2) / (count - 1)` but you have an equivalent of `(count - 1) / (count)`. I think that's the problem - one-off error.

Also, please address @iSazonov 's comments: remove the line with empty tag and  revert all the changes in `MeasureObjectStrings.resx` since we don't use this string for a custom error message anymore.

vors commented:
One more request: can you, please, add a new test for something like `1..100` (should be ~ `29.01149`), since we have a non-trivial algorithm it would be good to have more than 1 test for it.

CloudyDino committed:
Remove unused error AverageSwitchNotSet

CloudyDino committed:
Fixed calculation of standard deviation

CloudyDino committed:
Added test for standard deviation

Asks to calculate the standard deviation for the integers [1, 100]

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -708,8 +739,18 @@ private void AnalyzeString(string strValue, Statistics stat)
+            }
+            if (_measureStdDeviation && stat.count > 1)
+            {
+                double countMinusOne = stat.count - 1;
```
I don't think that introducing the `countMinusOne` variable makes the code cleaner. Let's just use `count - 1` inline instead. To force the `(count - 2) / (count -
 1)` to be a double division, instead of integer division, simply cast one of them to double, i.e.

```
(count - 2) / (count - 1.0)
```

iSazonov replied:
+1  I believe it is for compiler optimization.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -708,8 +739,18 @@ private void AnalyzeString(string strValue, Statistics stat)
+            {
+                double countMinusOne = stat.count - 1;
+                double avgPrevious = stat.sumPrevious / countMinusOne;
+                stat.variance *= (countMinusOne - 1)/countMinusOne;
```
Code style: add spaces around `/` like in other places.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -708,8 +739,18 @@ private void AnalyzeString(string strValue, Statistics stat)
                 stat.sum += numValue;
+            }
+            if (_measureStdDeviation && stat.count > 1)
+            {
```
let's leave a link in the comment to Wikipedia article the interative formula calculation - would be easier to reason in the future about this non-trivial part.

iSazonov replied:
I agree - I believe the public comment above is good place.

vors commented:
This is getting closer to be mergeble imo. Can somebody else take a quick look?

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1:
```diff
@@ -17,6 +18,25 @@ Describe "Measure-Object" -Tags "CI" {
+    It "Should calculate Standard Deviation" {
+        $actual = ($testObject | Measure-Object -Average -StdDeviation)
+        # We check this way since .StdDeviation returns a double value
+        [Math]::abs($actual.StdDeviation - 1.52752523165195) | Should -BeLessThan .00000000000001
```
We need to trust this. I think we need to add a comment from where we took  `1.52752523165195`. Below too.

CloudyDino replied:
Should we just add a comment that says we calculated it?

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Measure-Object.Tests.ps1:
```diff
@@ -17,6 +18,25 @@ Describe "Measure-Object" -Tags "CI" {
+    }
+
+    It "Should calculate Standard Deviation without -Average" {
+        $actual = ($testObject | Measure-Object -StdDeviation)
```
The test is duplicate previous test.
I believe we could chech that `-StdDeviation` works if other parameters present.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -71,6 +71,11 @@ public GenericMeasureInfo()
         public double? Minimum { get; set; }
+
+        /// <summary>
+        /// The Standard Deviation of property values
```
Please add final dot in all _public_ comments.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -835,6 +883,7 @@ private MeasureInfo CreateGenericMeasureInfo(Statistics stat, bool shouldUseGene
                 gmi.Count = stat.count;
                 gmi.Sum = sum;
                 gmi.Average = average;
+                gmi.StdDeviation = stdDeviation;
```
Is there a benefit if we put `stdDeviation` out too?

CloudyDino replied:
I'm not sure what you mean, can you elaborate?

CloudyDino replied:
Oh, I see what you mean; standard deviation doesn't seem to be as necessary. I don't think people would usually need it.

CloudyDino replied:
Currently, just deleting that line messes up the code. I'll try to understand it better and try again to work on this issue.

iSazonov commented:
@SteveL-MSFT Here we enhance a public structure - formally we need PowerShell Committee conclusion.

CloudyDino committed:
Update standard deviation tests

Checks to make sure that -StdDeviation works with -Sum and -Average

CloudyDino committed:
Update formatting and standard deviation calculation

Added periods to end of public comments and removed empty value tags as per iSazonov's request. Calculation of standard deviation is now cleaner.

CloudyDino committed:
Mention how values for standard deviation tests were calculated

CloudyDino committed:
Remove standard deviation from GenericMeasureInfo

iSazonov reviewed (commented):
Leave a comment

CloudyDino committed:
Unindent line

CloudyDino commented:
I realize that for some reason the tests started failing after commit c40ab47, I'm taking a look at what I did to see what went wrong.

CloudyDino committed:
revert commit 1820a41

Standard Devation wasn't being displayed after commit 1820a410507989e482f03649b5fd65b273f333d1

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/Measure-Object.cs:
```diff
@@ -128,6 +133,11 @@ public GenericObjectMeasureInfo()
+        /// <summary>
+        /// The Standard Deviation of property values.
+        /// </summary>
+        public double? StdDeviation { get; set; }
```
`StdDeviation` should be spelled out which is consistently with rest of PowerShell: `StandardDeviation`

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we are ok with adding a new parameter to this cmdlet.  The specific implementation details should be part of the regular PR review unless there is something controversial.

CloudyDino committed:
Change StdDeviation to StandardDeviation to be consistent

iSazonov commented:
@CloudyDino I wonder why you push the commit  1820a41?

CloudyDino commented:
@iSazonov I pushed 1820a41 on accident and then reverted it with commit df8ac60. I should have checked to see if everything still worked, but standard deviation wasn't being output due to that specific commit. Trying to figure out how to go about checking when we need to put `StandardDeviation` out.

SteveL-MSFT reviewed (approved):
LGTM.  Thanks for the contribution!

vors commented:
@iSazonov @CloudyDino thank you! 🎉

vors commented:
I filled https://github.com/PowerShell/PowerShell-Docs/issues/2183 for the help follow-up.

tylerl0706 commented:
So happy to see this get merged :) nice @CloudyDino!
